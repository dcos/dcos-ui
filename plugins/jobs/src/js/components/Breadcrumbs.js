import { Trans } from "@lingui/macro";
import { Link } from "react-router";
import PropTypes from "prop-types";
import React from "react";
import { ProductIcons } from "@dcos/ui-kit/dist/packages/icons/dist/product-icons-enum";

import Breadcrumb from "#SRC/js/components/Breadcrumb";
import BreadcrumbTextContent from "#SRC/js/components/BreadcrumbTextContent";
import PageHeaderBreadcrumbs from "#SRC/js/components/PageHeaderBreadcrumbs";

export default function Breadcrumbs({ jobPath, jobName, jobInfo, children }) {
  let breadcrumbParts = [
    <Breadcrumb key={"Jobs"} title="Jobs">
      <BreadcrumbTextContent>
        <Trans render={<Link to={"/jobs/overview/"} />}>Jobs</Trans>
      </BreadcrumbTextContent>
    </Breadcrumb>
  ];

  if (jobPath !== undefined) {
    const pathParts = jobPath.map((jobPathPart, index) => {
      const partId = jobPath.slice(0, index + 1).join(".");

      return (
        <Breadcrumb key={jobPathPart} title="Jobs">
          <BreadcrumbTextContent>
            <Link to={"/jobs/overview/" + partId}>{jobPathPart}</Link>
          </BreadcrumbTextContent>
        </Breadcrumb>
      );
    });
    breadcrumbParts = breadcrumbParts.concat(pathParts);
  }

  if (jobName !== undefined) {
    const partId = jobPath.concat([jobName]).join(".");
    breadcrumbParts = breadcrumbParts.concat([
      <Breadcrumb key={jobName} title="Jobs">
        <BreadcrumbTextContent>
          <Link to={"/jobs/detail/" + partId}>{jobName}</Link>
        </BreadcrumbTextContent>
        {jobInfo}
      </Breadcrumb>
    ]);
  }

  if (children !== undefined) {
    breadcrumbParts = breadcrumbParts.concat(React.Children.toArray(children));
  }

  return (
    <PageHeaderBreadcrumbs
      iconID={ProductIcons.Jobs}
      breadcrumbs={breadcrumbParts}
    />
  );
}

Breadcrumbs.propTypes = {
  jobInfo: PropTypes.node,
  jobName: PropTypes.string,
  jobPath: PropTypes.arrayOf(PropTypes.string),
  children: PropTypes.node
};
