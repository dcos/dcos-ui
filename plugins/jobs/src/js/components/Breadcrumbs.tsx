import { Trans } from "@lingui/react";
import { Link } from "react-router";
import * as React from "react";
import { ProductIcons } from "@dcos/ui-kit/dist/packages/icons/dist/product-icons-enum";

import Breadcrumb from "#SRC/js/components/Breadcrumb";
import BreadcrumbTextContent from "#SRC/js/components/BreadcrumbTextContent";
import PageHeaderBreadcrumbs from "#SRC/js/components/PageHeaderBreadcrumbs";

type Props = {
  jobInfo?: unknown;
  jobPath?: string[];
  jobName?: string;
  children?: React.ReactNode[];
};
export default ({ jobInfo, jobPath = [], jobName, children }: Props) => {
  let breadcrumbParts: React.ReactNode[] = [
    <Breadcrumb key={"Jobs"} title="Jobs">
      <BreadcrumbTextContent>
        <Trans render={<Link to={"/jobs/overview/"} />} id="Jobs" />
      </BreadcrumbTextContent>
    </Breadcrumb>,
  ];

  breadcrumbParts = breadcrumbParts.concat(
    jobPath.map((jobPathPart, index) => (
      <Breadcrumb key={jobPathPart} title="Jobs">
        <BreadcrumbTextContent>
          <Link to={"/jobs/overview/" + jobPath.slice(0, index + 1).join(".")}>
            {jobPathPart}
          </Link>
        </BreadcrumbTextContent>
      </Breadcrumb>
    ))
  );

  if (jobName !== undefined) {
    const partId = jobPath.concat([jobName]).join(".");
    breadcrumbParts = breadcrumbParts.concat([
      <Breadcrumb key={jobName} title="Jobs">
        <BreadcrumbTextContent>
          <Link to={"/jobs/detail/" + partId}>{jobName}</Link>
        </BreadcrumbTextContent>
        {jobInfo}
      </Breadcrumb>,
    ]);
  }

  if (children !== undefined) {
    breadcrumbParts = breadcrumbParts.concat(React.Children.toArray(children));
  }

  return (
    <PageHeaderBreadcrumbs
      iconID={ProductIcons.Jobs}
      breadcrumbs={breadcrumbParts}
    />
  );
};
