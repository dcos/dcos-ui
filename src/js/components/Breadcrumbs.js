import React, {PropTypes} from 'react';

import ManualBreadcrumbs from './ManualBreadcrumbs';

class Breadcrumbs extends React.Component {
  buildCrumbs(route) {
    let {router} = this.context;

    let crumbConfiguration = route.buildBreadCrumb(router);
    let crumbs = crumbConfiguration.getCrumbs(router);

    if (crumbConfiguration.parentCrumb) {
      let parentCrumb = router.getCurrentRoutes().find(function (eachRoute) {
        return eachRoute.path === crumbConfiguration.parentCrumb;
      });
      if (parentCrumb && parentCrumb.buildBreadCrumb) {
        crumbs = this.buildCrumbs(parentCrumb).concat(crumbs);
      }
    }

    return crumbs;
  }

  getCrumbsFromRoute() {
    let {router} = this.context;
    let routes = router.getCurrentRoutes();
    // Find the first route with a name
    let currentRoute = null;
    loop:
    for (var i = routes.length - 1; i >= 0; i--) {
      if (routes[i].buildBreadCrumb) {
        currentRoute = routes[i];
        break loop;
      }
    }

    if (!currentRoute) {
      return [];
    }

    let crumbs = this.buildCrumbs(currentRoute);

    return crumbs.slice(this.props.shift);
  }

  render() {
    let crumbs = this.getCrumbsFromRoute();

    return (
      <ManualBreadcrumbs crumbs={crumbs} />
    );
  }
};

Breadcrumbs.contextTypes = {
  router: PropTypes.func
};

Breadcrumbs.defaultProps = {
  // Remove root '/' by default
  shift: 0
};

Breadcrumbs.propTypes = {
  breadcrumbClasses: PropTypes.oneOfType([
    PropTypes.array,
    PropTypes.object,
    PropTypes.string
  ]),
  // Remove n crumbs from beginning of route crumbs
  shift: PropTypes.number
};

module.exports = Breadcrumbs;
