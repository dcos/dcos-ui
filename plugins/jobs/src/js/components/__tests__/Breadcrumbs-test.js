import React from "react";
import renderer from "react-test-renderer";

import Breadcrumbs from "../Breadcrumbs";

describe("Breadcrumbs", () => {
  const pathCases = [
    {
      name: "no path",
      path: []
    },
    {
      name: "a long path",
      path: ["foo", "bar"]
    }
  ];

  for (const pathCase of pathCases) {
    it(`has ${pathCase.name}`, () => {
      expect(
        renderer
          .create(
            <Breadcrumbs
              extra={<span>States</span>}
              path={pathCase.path}
              name="MyJob"
            />
          )
          .toJSON()
      ).toMatchSnapshot();
    });
  }

  it(`has path but no name`, () => {
    expect(
      renderer.create(<Breadcrumbs path={["foo", "bar"]} />).toJSON()
    ).toMatchSnapshot();
  });

  it(`has no props at all`, () => {
    expect(renderer.create(<Breadcrumbs />).toJSON()).toMatchSnapshot();
  });
});
