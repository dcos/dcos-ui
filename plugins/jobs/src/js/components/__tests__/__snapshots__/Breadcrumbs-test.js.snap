// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Breadcrumbs has a long path 1`] = `
<div
  className="breadcrumbs breadcrumbs--is-truncated"
>
  <div
    className="breadcrumb breadcrumb--is-icon"
  >
    <a
      onClick={[Function]}
      style={Object {}}
    >
      <svg
        aria-label="product-jobs icon"
        className="css-zgvqkl"
        height={24}
        preserveAspectRatio="xMinYMin meet"
        role="img"
        viewBox="0 0 24 24"
        width={24}
      >
        <use
          xlinkHref="#product-jobs"
        />
      </svg>
    </a>
  </div>
  <div
    className="breadcrumb breadcrumb--is-caret"
  >
    <svg
      aria-label="system-caret-right icon"
      className="css-1vunjz0"
      height={16}
      preserveAspectRatio="xMinYMin meet"
      role="img"
      viewBox="0 0 16 16"
      width={16}
    >
      <use
        xlinkHref="#system-caret-right"
      />
    </svg>
  </div>
  <div
    className="breadcrumb breadcrumb--force-ellipsis"
    onMouseEnter={[Function]}
    onMouseLeave={[Function]}
  >
    <div
      className="breadcrumb"
    >
      <div
        className="breadcrumb__content breadcrumb__content--text"
      >
        <a
          onClick={[Function]}
          style={Object {}}
        >
          foo
        </a>
      </div>
    </div>
  </div>
  <div
    className="breadcrumb breadcrumb--is-caret"
  >
    <svg
      aria-label="system-caret-right icon"
      className="css-1vunjz0"
      height={16}
      preserveAspectRatio="xMinYMin meet"
      role="img"
      viewBox="0 0 16 16"
      width={16}
    >
      <use
        xlinkHref="#system-caret-right"
      />
    </svg>
  </div>
  <div
    className="breadcrumb"
  >
    <div
      className="breadcrumb__content breadcrumb__content--text"
    >
      <a
        onClick={[Function]}
        style={Object {}}
      >
        bar
      </a>
    </div>
  </div>
  <div
    className="breadcrumb breadcrumb--is-caret"
  >
    <svg
      aria-label="system-caret-right icon"
      className="css-1vunjz0"
      height={16}
      preserveAspectRatio="xMinYMin meet"
      role="img"
      viewBox="0 0 16 16"
      width={16}
    >
      <use
        xlinkHref="#system-caret-right"
      />
    </svg>
  </div>
  <div
    className="breadcrumb"
  >
    <div
      className="breadcrumb__content breadcrumb__content--text"
    >
      <a
        onClick={[Function]}
        style={Object {}}
      >
        MyJob
      </a>
    </div>
    <span>
      States
    </span>
  </div>
</div>
`;

exports[`Breadcrumbs has no path 1`] = `
<div
  className="breadcrumbs"
>
  <div
    className="breadcrumb breadcrumb--is-icon"
  >
    <a
      onClick={[Function]}
      style={Object {}}
    >
      <svg
        aria-label="product-jobs icon"
        className="css-zgvqkl"
        height={24}
        preserveAspectRatio="xMinYMin meet"
        role="img"
        viewBox="0 0 24 24"
        width={24}
      >
        <use
          xlinkHref="#product-jobs"
        />
      </svg>
    </a>
  </div>
  <div
    className="breadcrumb breadcrumb--is-caret"
  >
    <svg
      aria-label="system-caret-right icon"
      className="css-1vunjz0"
      height={16}
      preserveAspectRatio="xMinYMin meet"
      role="img"
      viewBox="0 0 16 16"
      width={16}
    >
      <use
        xlinkHref="#system-caret-right"
      />
    </svg>
  </div>
  <div
    className="breadcrumb"
  >
    <div
      className="breadcrumb__content breadcrumb__content--text"
    >
      <a
        onClick={[Function]}
        style={Object {}}
      >
        Jobs
      </a>
    </div>
  </div>
  <div
    className="breadcrumb breadcrumb--is-caret"
  >
    <svg
      aria-label="system-caret-right icon"
      className="css-1vunjz0"
      height={16}
      preserveAspectRatio="xMinYMin meet"
      role="img"
      viewBox="0 0 16 16"
      width={16}
    >
      <use
        xlinkHref="#system-caret-right"
      />
    </svg>
  </div>
  <div
    className="breadcrumb"
  >
    <div
      className="breadcrumb__content breadcrumb__content--text"
    >
      <a
        onClick={[Function]}
        style={Object {}}
      >
        MyJob
      </a>
    </div>
    <span>
      States
    </span>
  </div>
</div>
`;

exports[`Breadcrumbs has no props at all 1`] = `
<div
  className="breadcrumbs"
>
  <div
    className="breadcrumb breadcrumb--is-icon"
  >
    <a
      onClick={[Function]}
      style={Object {}}
    >
      <svg
        aria-label="product-jobs icon"
        className="css-zgvqkl"
        height={24}
        preserveAspectRatio="xMinYMin meet"
        role="img"
        viewBox="0 0 24 24"
        width={24}
      >
        <use
          xlinkHref="#product-jobs"
        />
      </svg>
    </a>
  </div>
  <div
    className="breadcrumb breadcrumb--is-caret"
  >
    <svg
      aria-label="system-caret-right icon"
      className="css-1vunjz0"
      height={16}
      preserveAspectRatio="xMinYMin meet"
      role="img"
      viewBox="0 0 16 16"
      width={16}
    >
      <use
        xlinkHref="#system-caret-right"
      />
    </svg>
  </div>
  <div
    className="breadcrumb"
  >
    <div
      className="breadcrumb__content breadcrumb__content--text"
    >
      <a
        onClick={[Function]}
        style={Object {}}
      >
        Jobs
      </a>
    </div>
  </div>
</div>
`;

exports[`Breadcrumbs has path but no name 1`] = `
<div
  className="breadcrumbs"
>
  <div
    className="breadcrumb breadcrumb--is-icon"
  >
    <a
      onClick={[Function]}
      style={Object {}}
    >
      <svg
        aria-label="product-jobs icon"
        className="css-zgvqkl"
        height={24}
        preserveAspectRatio="xMinYMin meet"
        role="img"
        viewBox="0 0 24 24"
        width={24}
      >
        <use
          xlinkHref="#product-jobs"
        />
      </svg>
    </a>
  </div>
  <div
    className="breadcrumb breadcrumb--is-caret"
  >
    <svg
      aria-label="system-caret-right icon"
      className="css-1vunjz0"
      height={16}
      preserveAspectRatio="xMinYMin meet"
      role="img"
      viewBox="0 0 16 16"
      width={16}
    >
      <use
        xlinkHref="#system-caret-right"
      />
    </svg>
  </div>
  <div
    className="breadcrumb"
  >
    <div
      className="breadcrumb__content breadcrumb__content--text"
    >
      <a
        onClick={[Function]}
        style={Object {}}
      >
        Jobs
      </a>
    </div>
  </div>
  <div
    className="breadcrumb breadcrumb--is-caret"
  >
    <svg
      aria-label="system-caret-right icon"
      className="css-1vunjz0"
      height={16}
      preserveAspectRatio="xMinYMin meet"
      role="img"
      viewBox="0 0 16 16"
      width={16}
    >
      <use
        xlinkHref="#system-caret-right"
      />
    </svg>
  </div>
  <div
    className="breadcrumb"
  >
    <div
      className="breadcrumb__content breadcrumb__content--text"
    >
      <a
        onClick={[Function]}
        style={Object {}}
      >
        foo
      </a>
    </div>
  </div>
  <div
    className="breadcrumb breadcrumb--is-caret"
  >
    <svg
      aria-label="system-caret-right icon"
      className="css-1vunjz0"
      height={16}
      preserveAspectRatio="xMinYMin meet"
      role="img"
      viewBox="0 0 16 16"
      width={16}
    >
      <use
        xlinkHref="#system-caret-right"
      />
    </svg>
  </div>
  <div
    className="breadcrumb"
  >
    <div
      className="breadcrumb__content breadcrumb__content--text"
    >
      <a
        onClick={[Function]}
        style={Object {}}
      >
        bar
      </a>
    </div>
  </div>
</div>
`;
